### ISSUE TITLE
Gather keyboard input

### ISSUE THREAD

**Issue by Gavin-Williams** [microsoft/microsoft-ui-xaml#3986]
In UWP, this would be how I get keyboard input for a Xaml application ...

```cs
Microsoft.UI.Xaml.Window.Current.CoreWindow.KeyDown += CoreWindow_KeyDown;
```

This doesn't work for WinUI, with output... Microsoft.UI.Xaml.Window.Current.**get** returned null. How do we get keyboard input for the Window?

I should mention that gathering from a control, such as SwapChainPanel also doesn't work. It's a pretty weird design for sure.

**Comment by Gavin-Williams** [comment]
So I can get input from the SwapChainPanel, but not by any way that I would imagine. I can't get keyboard input from the SwapChainPanel itself, even though it exposes KeyUp and KeyDown events. I have to get input from a Grid that contains the SwapChainPanel. But the SCP is a Grid itself, but don't let that fool you, This Grid needs to be inside a another Grid - lol. This is a wacky design.

Can we please just have normal here and allow input from the Window, and allow input from the SwapChainPanel as well.

**Comment by mdtauk** [comment]
Doesn't the WinUI Window API spec mention being able to reference the CoreWindow for UWP, and the HWND for Win32 apps?

This is mentioned in the spec, but may not be implemented yet in preview 3

```
Window.CoreWindow property:

Get the CoreWindow of a Window when the Window was created in a UWP app model.

public CoreWindow CoreWindow { get; }

Remarks

Win32: This property returns null when running in Win32 app model.

```

**Comment by Gavin-Williams** [comment]
@mdtauk I can't understand why I shouldn't get the CoreWindow, whether I'm building a UWP.Desktop app or NonUWP.Desktop app, it's a modern window model, and should underlay all windowed applications. Shouldn't it? The name gives it away, it's a *CORE* Window. What's the thinking behind keeping around legacy models when they could be replaced with modern variants which are just as capable and use modern API styles and type systems? Are we expected to write redundant code to support both UWP and non-UWP scenarios?

Considering that we have UWP's CoreWindow and a XamlWindow. I don't see why I would want HWND in 2021?

I mean, getting keyboard input doesn't even make sense as it is, forcing us to use 20 year old API's side-by-side to get our applications to work isn't going to help is it?f

How does not returning a CoreWIndow help me?

**Comment by mdtauk** [comment]
> @mdtauk I can't understand why I shouldn't get the CoreWindow, whether I'm building a UWP.Desktop app or NonUWP.Desktop app, it's a modern window model, and should underlay all windowed applications. Shouldn't it? The name gives it away, it's a _CORE_ Window. What's the thinking behind keeping around legacy models when they could be replaced with modern variants which are just as capable and use modern API styles and type systems? Are we expected to write redundant code to support both UWP and non-UWP scenarios?
> 
> Considering that we have UWP's CoreWindow and a XamlWindow. I don't see why I would want HWND in 2021?
> 
> I mean, getting keyboard input doesn't even make sense as it is, forcing us to use 20 year old API's side-by-side to get our applications to work isn't going to help is it?
> 
> **How does not returning a CoreWindow help me?**

Win32 apps don't use a CoreWindow - it uses the normal win32 HWND interface,

CoreWindow has sandbox restrictions because it is for the UWP lifecycle.

WinUI's Window is a top level API that wraps both the HWND and CoreWindow, and re-directs to the appropriate one, depending on if you are in the UWP sandbox or not.

You can stay in the Window API level, but if you need to customise your window beyond what it provides, then there are ways to call into the CoreWindow if you stay in UWP or into the HWND if you are a win32 app.

If your app is UWP as you say it is, you should be able to call into the Window.CoreWindow and use it.  As a win32 app, this will return NULL.  if your app is in the UWP sandbox, then calling into HWND will not work.

**Comment by Gavin-Williams** [comment]
My app isn't UWP, but I'm using as much UWP code as possible, because really, all I want is UWP (with .Net 5 + NoOverlays + Steam support) which is only available via the desktop app model. If .Net 5 gets UWP and UWP get's Overlay suspension and .Net 5 compilation to get away from .Net native then I will likely go back to UWP. But we are stuck on what's being called Desktop - even though UWP is desktop as well.

CoreWindow doesn't have to have any restrictions - for example, overlays are not required to maintain the UWP security model.

And none of this helps with creating a sensible, consistent and modern way to get keyboard input from the Window or SwapChainPanel.

**Comment by mdtauk** [comment]
> My app isn't UWP, but I'm using as much UWP code as possible...
>
Well for the foreseeable, CoreWindow is off limits.  That is why it returns NULL

You will have to do it the Win32 way until more fundamental changes come, probably through Reunion APIs

**Comment by huoyaoyuan** [comment]
> From my perspective, Microsoft have introduced a new window model - the HWND - it's something that was used historically, many years ago, before my time, and they've brought it back, because _nostalgia_. It's weird because I've been using CoreWindow and I thought with .Net 5 we were getting a new unified App/Window model and framework. Instead, I have to dig into god knows where to get keyboard input, and I need to write redundant code to support this legacy thing.

Unifying them is the goal of Project Reunion, which is not yet ready now. It is on the roadmap.

**Comment by StephenLPeters** [comment]
@Austin-Lamb FYI

**Comment by Gavin-Williams** [comment]
OK. Thanks guys. I guess I was thinking that .Net 5 and WinUI were basically done. But maybe I'm jumping the gun a bit, and the picture isn't yet complete. I'll have to exercise some patience over the next year and wait and see what happens.

**Comment by saint4eva** [comment]
> > From my perspective, Microsoft have introduced a new window model - the HWND - it's something that was used historically, many years ago, before my time, and they've brought it back, because _nostalgia_. It's weird because I've been using CoreWindow and I thought with .Net 5 we were getting a new unified App/Window model and framework. Instead, I have to dig into god knows where to get keyboard input, and I need to write redundant code to support this legacy thing.
> 
> Unifying them is the goal of Project Reunion, which is not yet ready now. It is on the roadmap.

Could you explain this: cswinrt vs cswin32 vs project reunion?

**Comment by marb2000** [comment]
Thanks @huoyaoyuan and @mdtauk for helping with this.

@Gavin-Williams I see several gaps here, but these are the two that pop to me. One issue here is that the Window object is not a DependencyObject, so it does not have KeyDown / Up events. So you have to add a grid as content. The request could be like this: **The Window object should be a DependencyObject or similar.**

Another issue is that CoreDispatcher is useless in Win32, and we are losing the ability to **centralize keystrokes in a method using, for example, CoreDispatcher.AcceleratorKeyActivated.** The platform should give devs something that allows this. 
I do recommend opening the two requests. 

@saint4eva  About what is CSWinRT, CSWin32, abd Project Reunion:

- CswinRT or C#/WinRT is a tool that generates a .NET 5 projection libraries over WinRT. So you write C# code in a .NET 5 fashion way against the WinRT libraries (which is the type of libraries used by WinUI and UWP). 
- CSWin32 or C#/Win32 is the same concept but with Win32.  So you write C# code in a .NET 5 fashion way against the classic Win32 libraries.
- Project Reunion is a set of initiatives that will allow Win32 developers to use most pieces of the UWP platform (WinRT libraries and technologies). WinUI is part of it, as MRT Core, and other future components like Windowing, Win2D, etc.

**Comment by Gavin-Williams** [comment]
So, I thought I had keyboard input, but now I don't anymore. Has something changed? Why is this so difficult? I'm subscribing to both the grid and the swapchain for input. Neither trigger.

Here's my code..
`
    <Grid x:Name="XamlRoot" Background="Black" HorizontalAlignment="Stretch" VerticalAlignment="Stretch">
        <SwapChainPanel x:Name="XamlSwapChainPanel" x:FieldModifier="private" Width="Auto" Height="Auto"/>
    </Grid>
`
and here...
`    
    xamlRoot.KeyDown += XamlKeyDown;
    xamlRoot.KeyUp += XamlKeyUp;
    scp.KeyDown += XamlKeyDown;
    scp.KeyUp += XamlKeyUp;
`
  and this never gets called...
 `
   private void XamlKeyDown(object sender, KeyRoutedEventArgs e)
   {
       // this is never reached
   }
 `

**Comment by Gavin-Williams** [comment]
I tried this...
```xml
    <SwapChainPanel x:Name="XamlSwapChainPanel" x:FieldModifier="private" Width="Auto" Height="Auto">
        <Canvas x:Name="XamlCanvas" HorizontalAlignment="Stretch" VerticalAlignment="Stretch"/>
    </SwapChainPanel>
```
Because of this documentation...
```
The SwapChainPanel class does not inherit from the Control class, so you can't programmatically focus it directly for purposes of capturing key events. Consider setting the focus to a focusable element inside the panel and letting the key event bubble.
```
But that doesn't work either. The even is never called.

**Comment by Gavin-Williams** [comment]
I got it working. I think your documentation contradicts itself. On the surface, SCP doesn't collect input. but when I dug deeper, I found a comment...
```
For a control to receive input focus, it must be enabled, visible, and have IsTabStop and HitTestVisible property values of true.
```
So even though SCP is a Panel, not a control, I changed the values of IsTabStop and IsHitTestVisible to true and I get keyboard events firing on the SwapChain, supposedly something that's not meant to happen without bubbling from a child control.
